packages/compass-import-export/src/import/analyze-csv-fields.ts

```diff
@@ -0,0 +1,240 @@
+import type { Readable } from 'stream';
+import Papa from 'papaparse';
+
+import type { Delimiter } from '../utils/constants';
+import { createDebug } from '../utils/logger';
+import { csvHeaderNameToFieldName } from '../utils/csv-header';
+
+const debug = createDebug('analyze-csv-fields');
+
+const MIN_INT = -2147483648;
+const MAX_INT = 2147483647;
+const MIN_LONG = BigInt('-9223372036854775808');
+const MAX_LONG = BigInt('9223372036854775807');
+
+type AnalyzeCSVFieldsOptions = {
+  input: Readable;
+  delimiter: Delimiter;
+  abortSignal: AbortSignal;
+  progressCallback: (index: number) => void;
+};
+
+// see parseDynamic() in papaparse.js:
+// https://github.com/mholt/PapaParse/blob/aa0046865f1b4e817ebba6966d6baf483e0652d7/papaparse.js#L1225
+type PapaValue = boolean | number | Date | null | string;
+
+type PapaRowData = Record<string, PapaValue>;
+
+type CSVFieldTypeInfo = {
+  // How many cells in the file matched this type.
+  count: number;
+
+  // The line in the file where this field was first detected to be of this
+  // type. This is so the field type selector can immediately present that line
+  // or document as a counter example if the user selects an incompatible field
+  // type.
+  firstRowIndex: number;
+  firstColumnIndex: number;
+  firstValue: string;
+};
+
+// the subset of bson types that we can detect
+type CSVFieldType =
+  | 'int'
+  | 'long'
+  | 'double'
+  | 'boolean'
+  | 'date'
+  | 'string'
+  | 'null'
+  | 'mixed';
+
+/*
+For each field we need the detected types and the column positions. This helps
+with accounting for all column indexes, but also the fact that we'd have higher
+counts than the number of rows. ie. foo[0],foo[1] means twice as many fields as
+rows once it becomes field foo and so does foo[0].bar,foo[1].bar once it becomes
+foo.bar.
+*/
+type CSVField = {
+  types: Record<CSVFieldType, CSVFieldTypeInfo>;
+  columnIndexes: number[];
+  detected: CSVFieldType;
+};
+
+type AnalyzeCSVFieldsResult = {
+  totalRows: number;
+  fields: Record<string, CSVField>;
+};
+
+function detectFieldType(value: PapaValue): CSVFieldType {
+  // papaparse handles some types for us, so get those out of the way first
+  if (value === null) {
+    return 'null';
+  }
+  if (value instanceof Date) {
```

Comment:
This probably doesn’t break here, but I also see no reason to do the “real” brand check and use `Object.prototype.toString.call()` instead